### PR TITLE
fix(sqlite): fix column offset for join result

### PR DESCRIPTION
**Pull request by Hanaasagi**
### What does this PR do?

Close: #6597
Close: https://github.com/oven-sh/bun/issues/7147

When iterating the columns from `sqlite3_column_name`, we will get 

```
Id
Driver
Id
Name
```

It will contain duplicated column names But `PropertyNameArray` will not store them. The offset is mismatched.

|  `PropertyNameArray` | value from `sqlite3_column_value`  |
|---|---|
| Id  | 1 |
| Driver  | 1 |
| Name  | 1 |
|   | `Alice` |

This PR includes two kinds of fixes, I'm not sure which one to use.

- https://github.com/oven-sh/bun/commit/d56ed431b7c136c4e0ac2c3e48b8c8c1eda6389b  adds a bitmap `columnOffsets` to track the offsets corresponding to the names stored in `PropertyNameArray`. It will skip duplicated column name. If there are many columns, hashset might be more efficient. 

```
    Id: 1,
    Driver: 1,
    Name: "Alice",
```

- https://github.com/oven-sh/bun/commit/b77e6acea40dceeb5d236a94be5ac8ff56016591 just concat the table name and column name.

```
    Id: 1,
    Driver: 1,
    Users_Id: 1,
    Name: "Alice",
```

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?

I wrote automated tests.

### CODE REVIEW

File: src/bun.js/bindings/sqlite/JSSQLStatement.cpp [hunk 1]
```diff
@@ -265,7 +265,17 @@ static void initializeColumnNames(JSC::JSGlobalObject* lexicalGlobalObject, JSSQ
...
-            columnNames->add(Identifier::fromString(vm, WTF::String::fromUTF8(name, len)));
+            int preCount = columnNames->size();
+            columnNames->add(
+                Identifier::fromString(vm, WTF::String::fromUTF8(name, len))
+            );
+            int curCount = columnNames->size();
+
+            if (preCount == curCount) {
+                const char* tableName = sqlite3_column_table_name(stmt, i);
+                WTF::String fullName = WTF::String::fromUTF8(tableName, strlen(tableName)) + "_" + name;
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/sqlite/JSSQLStatement.cpp]:
```suggestion
                WTF::String fullName = makeString(tableName, "_"_s, name);
```

File: src/bun.js/bindings/sqlite/JSSQLStatement.cpp [hunk 2]
```diff
@@ -265,7 +265,17 @@ static void initializeColumnNames(JSC::JSGlobalObject* lexicalGlobalObject, JSSQ
...
             }
 
-            columnNames->add(Identifier::fromString(vm, WTF::String::fromUTF8(name, len)));
+            int preCount = columnNames->size();
+            columnNames->add(
+                Identifier::fromString(vm, WTF::String::fromUTF8(name, len))
+            );
+            int curCount = columnNames->size();
+
+            if (preCount == curCount) {
```
Comment (Jarred-Sumner) [inline src/bun.js/bindings/sqlite/JSSQLStatement.cpp]:
```suggestion
            // When joining multiple tables, the same column names can appear multiple times
            // columnNames de-dupes property names internally
            // We can't have two properties with the same name, so in that case we join the table name with the original name when there's a duplicate
            if (preCount == curCount) {
```

**Comment by Hanaasagi** [comment]
Btw, better-sqlite3 returns the last value of duplicate fields in above case.

Test code
```JavaScript
const db = require("better-sqlite3")();

db.prepare(
  "CREATE TABLE Users (Id INTEGER PRIMARY KEY, Name VARCHAR(255), CreatedAt TIMESTAMP)",
).run();
db.prepare(
  "CREATE TABLE Cars (Id INTEGER PRIMARY KEY, Driver INTEGER, CreatedAt TIMESTAMP, FOREIGN KEY (Driver) REFERENCES Users(Id))",
).run();
db.prepare("INSERT INTO Users (Id, Name, CreatedAt) VALUES (?, ?, ?);").run(
  1,
  "Alice",
  "2022-01-01",
);
db.prepare("INSERT INTO Cars (Id, Driver, CreatedAt) VALUES (?, ?, ?);").run(
  2,
  1,
  "2023-01-01",
);

const car = db
  .prepare("SELECT * FROM Cars JOIN Users ON Driver=Users.Id")
  .get();

console.log(car);

db.close();

```

Output:

```
{ Id: 1, Driver: 1, CreatedAt: '2022-01-01', Name: 'Alice' }
// `CreatedAt` is from table `Users`, not `Cars`.
```


Should we align this behavior with that of better-sqlite3 ?

**Comment by Jarred-Sumner** [comment]
I think using the last one is the most consistent with other database clients. I vaguely remember ActiveRecord being similar.

**Comment by Hanaasagi** [comment]
Close https://github.com/oven-sh/bun/issues/7147

**Comment by Hanaasagi** [comment]
@Jarred-Sumner  hi, any comments for this?

**Comment by Hanaasagi** [comment]
no response for a long time, closed.

**Pull request closed without merging**